### DISCUSSION TITLE
what is the difference between TxBuilder.bootstrap and TxBuilder?

### DISCUSSION THREAD

**Discussion by leverwwz** [input-output-hk/cardano-js-sdk#1458] [General]
1. following is TxBuilder.bootstrap test file 

https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/tx-construction/test/tx-builder/TxBuilder.bootstrap.test.ts

2. following is TxBuilder.test file 
 https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/tx-construction/test/tx-builder/TxBuilder.test.ts

3. what is the difference between TxBuilder.bootstrap and TxBuilder?

**Comment by mkazlauskas** [comment]
Hi, originally `TxBuilder.bootstrap.test.ts` was used to test the builder behavior when `addresses` provider resolves with empty `[]` array. I see that it's been incorrectly refactored since then and it has lost it's purpose.

So to answer your question - there's no difference, `bootstrap` needs to be either fixed to serve the original purpose, or deleted.

**Reply by leverwwz** [reply]
thanks.
but i am not sure that if use Byron address , we need to use BootstrapWitness?

can u help me please?    how to use VkeyWitness? or help to provide some demo.   
appreciate it so much.

**Reply by mkazlauskas** [reply]
If you want to use Byron address in an output then you don't need to do anything special.
If it's about spending funds owned by Byron address, then you're out of luck as we don't have any examples. There _might_ be some API issues or missing functionality.

>  we need to use BootstrapWitness?

Doesn't sound right

**Reply by leverwwz** [reply]
thank u so much

**Reply by leverwwz** [reply]
sir, any examples for spending funds of any type address?

create transaction and sign it.

**Reply by mkazlauskas** [reply]
Wallet e2e tests: https://github.com/input-output-hk/cardano-js-sdk/tree/master/packages/e2e/test
Dapp example (it's using some SDK functionality that's not yet published to npm): https://github.com/input-output-hk/SDET-dapp

**Reply by leverwwz** [reply]
thank u sir,

any example for offline?

**Reply by mkazlauskas** [reply]
No sorry. The 'online' part is the 'Providers' (where the data such as utxo comes from). You would have to implement your own.

**Reply by leverwwz** [reply]
got it,  thank you so much.
